### PR TITLE
Issue #2819 Add codecov.io into make tests and Travis

### LINKED ISSUE #2819: Add codecov.io into make tests and Travis
### General information

Add codecov.io integration into .travis.yml so that code coverage results are computed on each PR in travis and team would have additional check on PR to see whether a particular PR is not missing unit tests.

Additionally add test coverage generation into `make test` target - just adding `go test -cover` might be enough.

  * Minishift version: N/A
  * OS: N/A
  * Hypervisor: N/A

Some links:
* https://gist.github.com/jonfriesen/1c64bebdd2b30ba1b2465c259f070459
* https://github.com/codecov/example-go
* https://blog.golang.org/cover

### Steps to reproduce

  1. `make test`
  
### Expected

Code coverage results are generated.

### Actual

Code coverage results are not generated.

### PR DESCRIPTION
**Pull request by agajdosi**
Fixes #2819

Steps to test the pull request

  1. you can see it working on my fork: https://codecov.io/gh/agajdosi/minishift and https://travis-ci.org/agajdosi/minishift
  2. as admin of Minishift project you can enable Minishift project in codecov.io and then next build of this PR would probably upload the report

### CODE REVIEW

**Comment by agajdosi** [comment]
retest this please

**Comment by praveenkumar** [comment]
@agajdosi can you check https://docs.codecov.io/docs/error-reference#section-missing-base-commit and what it means missing the coverage?

**Comment by budhram** [comment]
retest this please

**Comment by agajdosi** [comment]
@praveenkumar The bot comes from my fork of Minishift and if I understand it correctly it compares the report of the branch with master on this fork which does not have any Codecov report yet. We need to enable codecov for this repository and once the code is merged and build for master is done, then we have the base coverage report which then will be compared with incoming branches.

**Comment by praveenkumar** [comment]
> The bot comes from my fork of Minishift and if I understand it correctly it compares the report of the branch with master on this fork which does not have any Codecov report yet.

Got you. So that mean we need to add our repo to codecov.

**Comment by agajdosi** [comment]
@praveenkumar Yes. It's quick and easy to do, however I do not have rights, so if you have you can try.

**Comment by praveenkumar** [comment]
@agajdosi I did check to codecov UI and looks like it is already added, not sure what else to do.

**Comment by agajdosi** [comment]
@praveenkumar I see. (As I do not have admin rights I did not saw minishift repo in my UI in codecov.io). From here https://codecov.io/gh/minishift/minishift it looks we are fine.

File: Makefile [hunk 1]
```diff
@@ -221,6 +221,11 @@ clean_bindata:
 test: $(ADDON_ASSET_FILE)  ## Run unit tests
 	@go test -v -tags "$(BUILD_TAGS_SYSTEMTRAY)" -ldflags="$(VERSION_VARIABLES)" $(shell $(PACKAGES))
 
+.PHONY: coverage
+coverage: $(ADDON_ASSET_FILE)
+	rm -f out/coverage.txt
+	@go test -v -tags "$(BUILD_TAGS_SYSTEMTRAY)" -ldflags="$(VERSION_VARIABLES)" -coverprofile=out/coverage.txt -covermode=atomic $(shell $(PACKAGES))
```
Comment (LalatenduMohanty) [inline Makefile]: I got below result in `out/coverage.txt` when I executed `make coverage`. I am not sure how to interpret the result here. 
```
mode: atomic
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:33.47,34.21 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:37.3,38.17 2 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:34.21,36.4 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:38.17,40.4 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:44.13,47.2 2 1
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:49.31,51.16 2 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:54.2,55.12 2 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:58.2,59.16 2 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:62.2,62.20 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:65.2,67.45 3 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:51.16,53.3 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:55.12,57.3 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:59.16,61.3 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/unset.go:62.20,64.3 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:33.56,35.25 2 4
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:41.2,41.21 1 4
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:35.25,37.17 2 4
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:37.17,39.4 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:41.21,43.3 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:43.8,45.3 1 4
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:48.48,49.33 1 5
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:54.2,54.70 1 2
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:49.33,50.21 1 177
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:50.21,52.4 1 3
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:59.74,62.2 2 1
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:64.71,66.16 2 3
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:69.2,70.12 2 3
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:66.16,68.3 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:73.72,75.16 2 1
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:78.2,79.12 2 1
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:75.16,77.3 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:82.73,84.15 2 1
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:89.2,90.12 2 1
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:84.15,85.45 1 1
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:85.45,87.4 1 2
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:93.58,98.16 4 2
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:107.2,107.12 1 2
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:98.16,100.3 1 2
github.com/minishift/minishift/cmd/minishift/cmd/config/util.go:100.8,106.3 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/view.go:49.47,51.13 2 0
github.com/minishift/minishift/cmd/minishift/cmd/config/view.go:54.3,55.17 2 0
github.com/minishift/minishift/cmd/minishift/cmd/config/view.go:58.3,59.61 2 0
github.com/minishift/minishift/cmd/minishift/cmd/config/view.go:51.13,53.4 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/view.go:55.17,57.4 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/view.go:59.61,61.4 1 0
github.com/minishift/minishift/cmd/minishift/cmd/config/view.go:66.13,74.2 5 1
xxxxxxxxxxxxxxxxx
```

**Comment by LalatenduMohanty** [comment]
@agajdosi I am having little hard time understanding the result, may we can have a discussion around this before merging this PR .

**Comment by budhram** [comment]
> I am having little hard time understanding the result, may we can have a discussion around this before merging this PR .

I completely agree here. This is something a small step in our working style.
This need to be discussed. Else, some might not even want it.

Like issue template.

**Comment by agajdosi** [comment]
> I am having little hard time understanding the result
@LalatenduMohanty AFAIU the resulting file is designated to be read by machines and parsed to more human readable formats, more info can be found here: https://stackoverflow.com/questions/31413281/golang-coverprofile-output-format. I would use the friendlier format but the issue is oriented toward codecov.io which requires this raw format of data.

> This need to be discussed. Else, some might not even want it.
@budhram As I understand it the goal here is not to introduce any hard requirements like we need to go to 100% of coverage or that 27% is bad. Instead the purpose of adding codecov.io is to give us more information about the quality of the code. If the overall coverage in the PR lowers significantly we will see it, will be alerted, we would be able to discuss and make a decision - if there are reasons for lowering the coverage, well then we can merge, if there are not then probably lets add the unit tests.

**Comment by LalatenduMohanty** [comment]
@agajdosi @minishift/minishift-dev I tried 
```
$ make coverage
$ cd out
$ go tool cover -html=coverage.txt -o coverage.html
```
And then I opened the html file in broswer. The report gives overall coverage for each `go` file but you have to select each file and check. It is not greatly useful but might be useful if someone put an effort to understand it. So I am ok to go ahead with the PR.  But we need to add this to the documentation.

**Comment by LalatenduMohanty** [comment]
> Instead the purpose of adding codecov.io is to give us more information about the quality of the code.

Minishift was already added to codecov.io before the this isn't it? Also, I do not understand how codecov.io is useful for us. I mean how to use it.

**Comment by agajdosi** [comment]
@LalatenduMohanty I am not sure here whether it is really needed to add documentation. Codecov reports seem to be pretty straight forward, for example: https://github.com/redhat-developer/odo/pull/741#issuecomment-423567757. Maybe we could just try it and see how and what could be documented, if anything.

**Comment by agajdosi** [comment]
Test comment.

**Comment by praveenkumar** [comment]
@agajdosi can you rebase it? @LalatenduMohanty I am OK with this PR without any doc as this mostly around the PR job side if we need some document then we can create a follow up issue.

**Comment by agajdosi** [comment]
@praveenkumar rebased

File: docs/source/contributing/developing.adoc [hunk 2]
```diff
@@ -185,6 +185,7 @@ For more {project} commands and flags, see the xref:../command-ref/minishift.ado
 === Unit Tests
 
 Unit tests run on *Travis* before the code is merged.
+Unit test coverage is then reported to *Codecov* which provides coverage information on the pull requests.
```
Comment (LalatenduMohanty) [inline docs/source/contributing/developing.adoc]: As discussed with @agajdosi he is going to add steps to generate the report manually as mentioned in https://github.com/minishift/minishift/pull/2844#issuecomment-427718017

suggestion:
Locally you can generate a HTML code coverage  report by following steps:
<Steps>

**Comment by agajdosi** [comment]
@LalatenduMohanty suggested changes implemented, you can take a look

**Comment by agajdosi** [comment]
retest this please

**Comment by LalatenduMohanty** [comment]
retest this please

**Pull request merged**